### PR TITLE
#4 add basic clinical rev output

### PR DESCRIPTION
**Pull request by AnilMistry**
Towards #4

### CODE REVIEW

**Comment by tomaroberts** [comment]
Thanks Anil – screenshot looks great. I will review asap.

NB: the GitHub Action is failing because TotalSegmentator calls `sklearn` as a package, which is a deprecated term as it should be `scikit-learn`. I've got a separate PR which sorts this out.

File: app/operators/clinrev_pdf_operator.py [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+from reportlab.lib.units import inch, cm
+from reportlab.lib import utils
+
+from reportlab.lib import colors
+from reportlab.lib.colors import white, black
+from reportlab.lib.fonts import tt2ps
+from reportlab.rl_config import canvas_basefontname as _baseFontName
+_baseFontNameB = tt2ps(_baseFontName, 1, 0)
+_baseFontNameI = tt2ps(_baseFontName, 0, 1)
+_baseFontNameBI = tt2ps(_baseFontName, 1, 1)
```
Comment (tomaroberts) [inline app/operators/clinrev_pdf_operator.py:40]: There are quite a lot of unused imports, which you can remove. They will be greyed out in PyCharm.

File: app/operators/clinrev_pdf_operator.py [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+
+    """Generates a PDF with Sag/Cor/Ax views for each structure"""
+
+    def compute(self, op_input: InputContext, op_output: OutputContext, context: ExecutionContext):
+
+        logging.info(f"Begin {self.compute.__name__}")
+        # get list of masks
+        input_path = op_input.get("input_files").path
+        dcm_input_path = input_path / 'dcm_input'
+        nii_seg_output_path = dcm_input_path / 'nii_seg_output'
```
Comment (tomaroberts) [inline app/operators/clinrev_pdf_operator.py:57]: To get it working on my computer with a set of DICOM files in the `/input` directory, I had to change this line to:

`nii_seg_output_path = input_path / 'nii_seg_output'`

File: app/operators/clinrev_pdf_operator.py [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+                                                sag_img_path=sag_img_path,
+                                                cor_img_path=cor_img_path,
+                                                nii_filenames=nii_filenames)
+
+        logging.info(f"DICOM Encapsulated PDF written to {pdf_filename}")
+        logging.info(f"PDF creation complete ...")
+        logging.info(f"End {self.compute.__name__}")
+        op_output.set(value=DataPath(pdf_filename), label='pdf_file')
+
+        return
```
Comment (tomaroberts) [inline app/operators/clinrev_pdf_operator.py:88]: I don't think the `compute()` functions within an Operator class are supposed to have `return` statements.

File: app/operators/dcm2nii_operator.py [hunk 4]
```diff
@@ -40,11 +39,16 @@ def compute(self, op_input: InputContext, op_output: OutputContext, context: Exe
...
             shutil.move(os.path.join(input_path, f), dcm_input_path)
 
+        # create output directory for input-ct-dataset.nii.gz
+        dcm_output_path = os.path.join(input_path, 'nii_ct_output')
+        if not os.path.exists(dcm_output_path):
+            os.makedirs(dcm_output_path)
+
         # Run dcm2niix
-        subprocess.run(["dcm2niix", "-z", "y", "-o", input_path, "-f", "input-ct-dataset", dcm_input_path])
+        # subprocess.run(["dcm2niix", "-z", "y", "-o", dcm_output_path, "-f", "input-ct-dataset", dcm_output_path])
```
Comment (tomaroberts) [inline app/operators/dcm2nii_operator.py:48]: This should not be commented out for production.

File: app/operators/dcm2nii_operator.py [hunk 5]
```diff
@@ -40,11 +39,16 @@ def compute(self, op_input: InputContext, op_output: OutputContext, context: Exe
...
 
         # Delete superfluous .json files
-        json_files = glob.glob(input_path + "/*.json")
+        json_files = glob.glob(dcm_output_path + "/*.json")
```
Comment (tomaroberts) [inline app/operators/dcm2nii_operator.py:51]: I notice you've changed some of the paths, which might explain my other comment saying I couldn't get it working without changing a line 57 in the clinrev_operator.py.

I think we should get your PR working and then we'll do a follow-up to make the pathing more sensible in the MAP. It's a worthwhile improvement in the long run.

File: app/operators/totalsegmentator_operator.py [hunk 6]
```diff
@@ -34,7 +34,7 @@ def compute(self, op_input: InputContext, op_output: OutputContext, context: Exe
             os.makedirs(nii_seg_output_path)
 
         # Run TotalSegmentator
-        subprocess.run(["TotalSegmentator", "-i", nii_input_file, "-o", nii_seg_output_path])
+        # subprocess.run(["TotalSegmentator", "-i", nii_input_file, "-o", nii_seg_output_path])
```
Comment (tomaroberts) [inline app/operators/totalsegmentator_operator.py:37]: This should not be commented out for production MAP.

File: requirements.txt
Comment (tomaroberts) [inline requirements.txt:1]: Can you specify versions for these new packages please.

**Review by tomaroberts — CHANGES_REQUESTED** [review]
Thanks for the Anil – the PDF looks great.

I haven't fully finished going through it, but some initial comments with this review.

Broad comments:
- There are quite a lot of unused import statements across the files – can you remove these please. They will be greyed out in PyCharm.
- What is the `notebook` folder for? Is that meant to be a Jupyter notebook? If so, the uploaded file is a `.py` rather than `.ipynb`
- General flake8 formatting issues, but we can sort those once PR is working as we want.

File: requirements.txt
Comment (AnilMistry) [reply requirements.txt:1]: Added

File: app/operators/totalsegmentator_operator.py (hunk 6 above)
Comment (AnilMistry) [reply app/operators/totalsegmentator_operator.py]: un-uncommented

File: app/operators/dcm2nii_operator.py (hunk 4 above)
Comment (AnilMistry) [reply app/operators/dcm2nii_operator.py]: un-uncommented

File: app/operators/clinrev_pdf_operator.py (hunk 3 above)
Comment (AnilMistry) [reply app/operators/clinrev_pdf_operator.py]: Removed

File: app/operators/clinrev_pdf_operator.py (hunk 1 above)
Comment (AnilMistry) [reply app/operators/clinrev_pdf_operator.py]: Removed

**Comment by AnilMistry** [comment]
Updates made as requested. This works locally for me. Instead of having the ct niftis and masks in the input folder, I have them in a separated directory in the repo root, and pull them in when needed, after commenting out the subprocesses.

**Comment by AnilMistry** [comment]
I have update the paths so that it uses each operators' working directory to place their outputs, then assign the datapaths as the output. This works locally with premade Nifti files for th CT and Segments.

**Comment by tomaroberts** [comment]
Made some changes to ensure it works with a folder of DICOMs. Had to correct some path issues. Also, I discovered that the PDF creation fails if one of the DICOM elements required to fill in the PDF is missing, so I've added a little function to check for this and simply output "Unknown" if it doesn't exists.

TODO:
- create new MAP and run end-to-end on DGX/headnode
- run MAP on AIDE and check PDF functions as expected

**Comment by tomaroberts** [comment]
Tested latest commit on DGX (nb: remember to change line 26 in Dockerfile to this branch name) and it successfully ran end-to-end (DICOM in/out). Next to test on AIDE.

**Comment by tomaroberts** [comment]
#27 should be resolved by [0da0875](https://github.com/GSTT-CSC/TotalSegmentator-AIDE/pull/21/commits/0da087550071cd016b02cadea7e703d7f7d88a0a) – confirm in testing

**Comment by AnilMistry** [comment]
Ready for review, PDF should look like this: 
image

**Comment by tomaroberts** [comment]
Confirmed working on AIDE (with a few tweaks to AIDE code) by Joe B:

image

TODO: merge, create final 0.2.0 container, test on AIDE PreProd, release on GH, release to AIDE App Store.

**Review by tomaroberts — APPROVED** [review]
Going to merge. Thanks Anil – this was a great effort :)
